### ISSUE TITLE
Org-roam-capture templates are not working as expected

### ISSUE THREAD

**Issue by myshevchuk** [jethrokuan/org-roam#285]
First of all, thank you very much for your idea to create org-roam and for the work already done!
I realize that the transition is being made from a home-made capture mechanism to org-capture and many things are not yet supposed to work as intended. So this is rather a reference and workaround for those who need the templating power of org-caputre.

### Description
The templating system of org-capture is not used.
#### Steps to Reproduce

1. set `org-roam-capture-templates` to 
```
("d" "default" plain (function org-roam--capture-get-point)
     "%?Hello world!"
     :file-name "%<%Y%m%d%H%M%S>-${slug}"
     :head "#+TITLE: ${title}"
     :unnarrowed t)
```
2. Run `org-roam-find-file`
3. Create a file with title "The Title"

#### Expected Results
A new file is created and its buffer is visited. The contents of the file is 

>#+TITLE: The Title
>Hello world!

The point is moved to the position before letter "H"

#### Actual Results
A new file is created and its buffer is visited. The contents of the file is 
>#+TITLE: The Title

The org-capture template part 
>Hello world! 

Is missing

### Versions
- Emacs (`C-h v emacs-version`): v26.3
- Org-roam commit: https://github.com/jethrokuan/org-roam/commit/0cab66

**Comment by myshevchuk** [comment]
### Workaround 
Patch the function `org-roam--capture-new-file` by replacing:
```
(org-roam--fill-template (or (org-capture-get :head)
                                  org-roam--capture-header-default)
                              org-roam--capture-info)
```

with
```
(org-roam--fill-template (concat (or (org-capture-get :head)
                                              org-roam--capture-header-default) "\n"
                                          (org-capture-get :template))
                                  org-roam--capture-info)
```
Please note that the `%?` template key doesn't work with this workaround, instead verbatim `%?` is inserted.
What works: plain text, `%^{PROMPT}`

### Suggestions
1. Use org-capture-templates custom file path option `(keys description **target** template properties)`, in this case `(function org-roam--capture-get-point)`, as intended by `org-capture` design, that is return the point to the file in question. Creating a new file is fine in this case but do not write anything to a buffer.
2. Delegate all the templating functions to the template part of a capture template, i.e. fourth member of the template list `(keys description target **template** properties)`
3. Org-roam extends org-capture by injecting a custom header. Maybe it could be done by injecting the desired text using `org-capture-put`, which seems to be a fairly general mechanism, inside the wrapper function `org-roam-capture`.

**Comment by jethrokuan** [comment]
Thanks for the detailed report. That's the nature of `org-roam-find-file`, you'll notice that if files are created using `org-roam-insert` the behaviour is different.

The best way to workaround this is to include all of your template content in `:head`. Even in `org-roam-insert`, the capture template's original template is inserted only once, so they serve the same purpose.

your workaround would destroy template insertion for `org-roam-insert`.

**Comment by myshevchuk** [comment]
Ok, I see. I did not check it with `org-roam-insert` yet. Unfortunately, including all of the template in `:head` is not the best option. I landed here after trying to incorporate the following into my workflow:
```
:head "#+TITLE: ${title}\n
#+ROAM_KEY: cite:${citekey}\n
\n
fullcite:${citekey}"
```
The above: 
1. Asks twice for the `${citekey}` variable 
2. There is no means to set the point

The above is trivial with org-capture:
```
"#+ROAM_KEY: cite:%^{PROMPT}\n
%?\n
fullcite:%\1"
```
Not a deal-breaker for org-roam of course and surely this can be easily implemented in org-roam, but honestly, your package will only benefit if you make it as compatible and transparent with regard to the existing org-mode infrastructure and stick to the DRY principle as much as possible.

**Comment by myshevchuk** [comment]
With `org-roam-insert` and ***unmodified*** `org-roam--capture-new-file` I get the following error:

>org-capture: Capture template ‘r’: Invalid search bound (wrong side of point)

the value of `org-roam-capture-templates` is:
```
'(("d" "default" plain (function org-roam--capture-get-point)
          "%?"
           :file-name "%<%Y-%m-%d-%H%M%S>-${slug}"
           :head "#+TITLE: ${title} "
           :unnarrowed t)
  ("r" "reference" plain (function org-roam--capture-get-point)
          "User input: %^{PROMPT}"
          :file-name "references/%<%Y-%m-%d-%H%M%S>-${slug}"
          :head "#+TITLE: ${title}"
          :unnarrowed t)))
```

**Comment by jethrokuan** [comment]
> surely this can be easily implemented in org-roam

Well I've been thinking about how to do it for a while and haven't come to a solution, so if you can think of one do let me know. Maybe it's actually easy and I'm missing something.

There basically needs to be 2 different templates. 1 for insertion only once ever, and 1 for any new incoming content, as per normal org-capture.

**Comment by myshevchuk** [comment]
I could look into it but could you please elaborate on this:

> 2 different templates. 1 for insertion only once ever, and 1 for any new incoming content

**Comment by jethrokuan** [comment]
> With `org-roam-insert` and _**unmodified**_ `org-roam--capture-new-file` I get the following error:
> 
> > org-capture: Capture template ‘r’: Invalid search bound (wrong side of point)
> 
> the value of `org-roam-capture-templates` is:
> 
> ```
> '(("d" "default" plain (function org-roam--capture-get-point)
>           "%?"
>            :file-name "%<%Y-%m-%d-%H%M%S>-${slug}"
>            :head "#+TITLE: ${title} "
>            :unnarrowed t)
>   ("r" "reference" plain (function org-roam--capture-get-point)
>           "User input: %^{PROMPT}"
>           :file-name "references/%<%Y-%m-%d-%H%M%S>-${slug}"
>           :head "#+TITLE: ${title}"
>           :unnarrowed t)))
> ```

This seems to be a bug upstream. Trying:

```
(setq org-capture-templates '(("r" "reference" plain (file "~/inbox.org")
                               "User input: %^{PROMPT}"
                               :unnarrowed t)))
```

and a regular `org-capture`, I get the same error, which is thrown on:

```
(search-forward "%?" end t)
```

where `t` stands for NOERROR, seems like this arg is somehow not respected (i'm on Emacs 28).

**Comment by jethrokuan** [comment]
Your original template now works as expected.

**Comment by real-or-random** [comment]
> Ok, I see. I did not check it with `org-roam-insert` yet. Unfortunately, including all of the template in `:head` is not the best option. I landed here after trying to incorporate the following into my workflow:
> 
> ```
> :head "#+TITLE: ${title}\n
> #+ROAM_KEY: cite:${citekey}\n
> \n
> fullcite:${citekey}"
> ```
> 

Am I right that something like this still doesn't work nicely? I'd like to have the same `${property}` in `:head` and in `:file-name` but I'll ask twice for it. (And `%\1` doesn't work.)
